release-23.1: cli/demo: various fixes

knz opened the pull request:
Backport 6/6 commits from #108566.

/cc @cockroachdb/release

---

See individual commits for details.

Fixes #96239.
Fixes #102257.
Fixes #108331.
Fixes #108563.
Fixes #107888.

Epic: CRDB-28893

---
Release justification: removes test flakes

knz committed:
cli/demo: fix `\demo add` with multi-tenant clusters

Prior to this fix, secondary tenant servers were only started when the
demo cluster was initialized. Subsequent node restarts did not also
restart the secondary tenant server. This caused any
operation (including `\demo ls`) that access the secondary tenant to
fail.

This patch ensures that the secondary tenant server is started every
time a demo node is restarted.

Release note: None

knz committed:
cli/demo: factor some code

Release note: None

knz committed:
cli/demo: fix NumNodes

The `NumNodes()` method is meant to return the number of nodes. So it
can vary as demo nodes are shut down or restarted.

Given where it is called in `clisqlshell`, there is no user-visible
change here. However, the intent is different to `NumServers` which
does the other thing.

Release note: None

knz committed:
cli/demo: fix a crash when adding new servers

The `GetLocality()` method takes a node ID and was incorrectly using
that to index the server array directly. This is incorrect - the node
IDs are properties of the server and a lookup must be performed to
find the server.

Release note (bug fix): A bug was fixed in `cockroach demo` whereby
`\demo add` could sometimes crash with an error "`index out of range
[...] with length ...`". This bug had been introduced in v19.x.

knz committed:
cli/interactive_tests: unskip `test_demo_multitenant.tcl`

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/108631)
<!-- Reviewable:end -->

knz committed:
cli/demo: remove `\demo recommission`

Release note (bug fix): A bug was fixed whereby the command `\demo
decommission` in `cockroach demo` could sometime leave the demo
cluster in a broken state. This bug had been introduced in v20.2.

Release note (cli change): The command `\demo recommission` has been
removed from `cockroach demo`. It had been obsolete and non-functional
ever since v20.2.

rafiss reviewed (approved):
tyvm!
